arangod/Aql/DocumentProducingNode.cpp

```diff
@@ -174,3 +172,38 @@ void DocumentProducingNode::setFilterProjections(aql::Projections projections) {
 bool DocumentProducingNode::doCount() const noexcept {
   return _count && !hasFilter();
 }
+
+void DocumentProducingNode::recalculateProjections(ExecutionPlan* plan) {
+  _filterProjections.clear();
+  _projections.clear();
+  LOG_DEVEL << __PRETTY_FUNCTION__;
+  containers::FlatHashSet<AttributeNamePath> attributes;
+  if (hasFilter()) {
+    if (Ast::getReferencedAttributesRecursive(
+            this->filter()->node(), this->outVariable(),
+            /*expectedAttribute*/ "", attributes,
+            plan->getAst()->query().resourceMonitor())) {
+      _filterProjections = aql::Projections{std::move(attributes)};
+      LOG_DEVEL << "found filter projections: " << _filterProjections;
+    }
+  }
+
+  attributes.clear();
+  if (utils::findProjections(dynamic_cast<ExecutionNode*>(this), outVariable(),
+                             /*expectedAttribute*/ "",
+                             /*excludeStartNodeFilterCondition*/ true,
+                             attributes)) {
+    if (attributes.size() <= maxProjections()) {
+      LOG_DEVEL << "found projections: " << _projections;
+      _projections = Projections(std::move(attributes));
+    } else {
+      LOG_DEVEL << "too many projections: " << attributes << " ("
+                << attributes.size() << ") max: " << maxProjections();
+    }
+  }
+}
+
+bool DocumentProducingNode::isProduceResult() const {
```

Comment:
Should this be a pure virtual function?
Asking because `IndexNode::isProduceResult()` does not call the base class `DocumentProducingNode::isProduceResult()`. 
And the only other implementation is in `EnumerateCollectionNode::isProduceResult()`. So we could move the default implementation from `DocumentProducingNode` into `EnumerateCollectionNode`.